lib/spack/spack/cmd/mirror.py

```diff
@@ -231,142 +234,225 @@ def mirror_list(args):
+
 
-    verb = "updated" if existed else "created"
+def process_mirror_stats(present, mirrored, error):
+    p, m, e = len(present), len(mirrored), len(error)
     tty.msg(
-        "Successfully %s mirror in %s" % (verb, directory),
         "Archive stats:",
         "  %-4d already present" % p,
         "  %-4d added" % m,
         "  %-4d failed to fetch." % e,
     )
     if error:
         tty.error("Failed downloads:")
-        tty.colify(s.cformat("{name}{@version}") for s in error)
+        colify.colify(s.cformat("{name}{@version}") for s in error)
         sys.exit(1)
 
 
+def mirror_directory_from_cli(args):
+    mirror_path = args.directory or spack.util.path.canonicalize_path(
+        spack.config.get("config:source_cache")
+    )
+    mirror = spack.mirror.Mirror(mirror_path)
+    directory = url_util.format(mirror.push_url)
+    return directory
+
+
+def mirror_create(args):
+    """Create a directory to be used as a spack mirror, and fill it with
+    package archives.
+    """
+    if args.specs and args.all:
+        raise SpackError(
+            "cannot specify specs on command line if you chose to mirror all specs with '--all'"
+        )
+
+    if args.file and args.all:
+        raise SpackError(
+            "cannot specify specs with a file if you chose to mirror all specs with '--all'"
+        )
+
+    if args.file and args.specs:
+        raise SpackError("cannot specify specs with a file AND on command line")
+
+    if not args.specs and not args.file and not args.all:
+        raise SpackError(
+            "no packages were specified.",
+            "To mirror all packages, use the '--all' option "
+            "(this will require significant time and space).",
+        )
+
+    if args.versions_per_spec and args.all:
+        raise SpackError(
+            "cannot specify '--versions_per-spec' and '--all' together",
+            "The option '--all' already implies mirroring all versions for each package.",
+        )
+
+    if args.all and not ev.active_environment():
+        create_mirror_for_all_specs(args)
+        return
+
+    if args.all and ev.active_environment():
+        create_mirror_for_all_specs_inside_environment(args)
+        return
+
+    create_mirror_for_individual_specs(args)
+
+
+def create_mirror_for_all_specs(args):
+    mirror_specs = all_specs_with_all_versions(args)
+    directory = mirror_directory_from_cli(args)
+    mirror_cache, mirror_stats = spack.mirror.mirror_cache_and_stats(
+        directory, skip_unstable_versions=args.skip_unstable_versions
+    )
+    for candidate in mirror_specs:
+        pkg_cls = spack.repo.path.get_pkg_class(candidate.name)
+        pkg_obj = pkg_cls(spack.spec.Spec(candidate))
+        mirror_stats.next_spec(pkg_obj.spec)
+        spack.mirror.create_mirror_from_package_object(pkg_obj, mirror_cache, mirror_stats)
+    process_mirror_stats(*mirror_stats.stats())
+
+
+def create_mirror_for_all_specs_inside_environment(args):
```

Comment:
> For example I think the only thing this needs is the excluded specs

It needs also the `directory` where to place the mirror and whether or not to skip unstable versions. 

> Would it be possible to resolve args into the actual arguments used by each function at this level?

Done, though I have to say I slightly prefer the previous version of the code.[^1]  

[^1]: I prefer the previous version since the main `mirror_create` function was just doing error checking and then forwarding to other function if `args` passed all checks - so a terser view of the overall behavior without further details expanded. It's a minor style preference though.

Reply:
Thanks I appreciate this change (I often get confused when `args` from commands is passed transitively, since for me the method signature helps me understand at-a-glance what the function is "planning" to do, and when the parameters are grouped into a single opaque object I have to read the whole function and possibly function subcalls to understand it).